File: frontend/lib/insights.ts
import { api } from "@/lib/api-client";

// ---------------------------------------------------------------- vocabularies

export type OpportunityType =
  | "low_ctr_high_impressions"
  | "striking_distance"
  | "rising_query"
  | "content_gap"
  | "decaying_content"
  | "cannibalization";

export type OpportunityStatus = "open" | "in_progress" | "actioned" | "dismissed" | "expired";

export type AlertType =
  | "traffic_drop"
  | "ranking_drop"
  | "ctr_drop"
  | "content_decay"
  | "gsc_sync_failure"
  | "audit_score_drop"
  | "indexing_issue";

export type AlertSeverity = "info" | "warning" | "critical";

export type AlertStatus = "active" | "acknowledged" | "resolved" | "muted";

export const OPPORTUNITY_TYPE_LABELS_FA: Record<string, string> = {
  low_ctr_high_impressions: "نرخ کلیک پایین با نمایش بالا",
  striking_distance: "نزدیک به صفحه اول",
  rising_query: "عبارت جستجوی رو به رشد",
  content_gap: "خالی محتوایی",
  decaying_content: "افت تدریجی محتوا",
  cannibalization: "تداخل محتوایی (Cannibalization)",
};

export const OPPORTUNITY_STATUS_LABELS_FA: Record<string, string> = {
  open: "باز",
  in_progress: "در حال بررسی",
  actioned: "اقدام‌شده",
  dismissed: "رد شده",
  expired: "منقضی‌شده",
};

export const ALERT_TYPE_LABELS_FA: Record<string, string> = {
  traffic_drop: "افت ترافیک",
  ranking_drop: "افت رتبه",
  ctr_drop: "افت نرخ کلیک",
  content_decay: "افت تدریجی محتوا",
  gsc_sync_failure: "خطای همگام‌سازی سرچ کنسول",
  audit_score_drop: "افت امتیاز حسابرسی",
  indexing_issue: "مشکل ایندکس شدن",
};

export const ALERT_SEVERITY_LABELS_FA: Record<string, string> = {
  info: "اطلاعات",
  warning: "هشدار",
  critical: "بحرانی",
};

export const ALERT_STATUS_LABELS_FA: Record<string, string> = {
  active: "فعال",
  acknowledged: "تایید شده",
  resolved: "برطرف شده",
  muted: "بی‌صدا شده",
};

// -------------------------------------------------------------------- entities

export interface DashboardWebsiteRow {
  website_id: string;
  name: string;
  domain: string;
  clicks: number;
  impressions: number;
  ctr: number;
  avg_position: number;
  health_score: number;
  open_alerts: number;
  open_opportunities: number;
}

export interface DashboardSummary {
  organization_id: string;
  website_count: number;
  project_count: number;
  health_score: number;
  total_clicks: number;
  total_impressions: number;
  avg_ctr: number;
  avg_position: number;
  clicks_change_percent: number | null;
  impressions_change_percent: number | null;
  active_alerts: number;
  critical_alerts: number;
  open_opportunities: number;
  estimated_traffic_gain: number;
  published_articles: number;
  draft_articles: number;
  last_audit_score: number | null;
  last_gsc_sync_at: string | null;
  websites: DashboardWebsiteRow[];
}

export interface Opportunity {
  id: string;
  organization_id: string;
  website_id: string;
  opportunity_type: OpportunityType | string;
  status: OpportunityStatus | string;
  title: string;
  description: string | null;
  query: string | null;
  page_url: string | null;
  keyword_id: string | null;
  priority_score: number;
  estimated_traffic_gain: number;
  current_position: number | null;
  current_clicks: number;
  current_impressions: number;
  current_ctr: number | null;
  details: Record<string, any>;
  recommended_action: string | null;
  detected_at: string;
  last_seen_at: string | null;
  actioned_at: string | null;
  dismissed_at: string | null;
  dismiss_reason: string | null;
  linked_brief_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface OpportunitySummary {
  total_open: number;
  by_type: Record<string, number>;
  total_estimated_traffic_gain: number;
  top: Opportunity[];
}

export interface OpportunityDetectResult {
  website_id: string;
  scanned_queries: number;
  scanned_pages: number;
  created: number;
  updated: number;
  expired: number;
  by_type: Record<string, number>;
}

export interface Alert {
  id: string;
  organization_id: string;
  website_id: string;
  alert_type: AlertType | string;
  severity: AlertSeverity | string;
  status: AlertStatus | string;
  title: string;
  message: string;
  metric_name: string | null;
  current_value: number | null;
  previous_value: number | null;
  change_percent: number | null;
  entity_type: string | null;
  entity_id: string | null;
  details: Record<string, any>;
  occurrence_count: number;
  triggered_at: string;
  last_seen_at: string | null;
  acknowledged_at: string | null;
  resolved_at: string | null;
  resolution_note: string | null;
  muted_until: string | null;
  notified_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface AlertSummary {
  active: number;
  critical: number;
  warning: number;
  info: number;
  by_severity: Record<string, number>;
}

export interface Notification {
  id: string;
  organization_id: string;
  user_id: string | null;
  website_id: string | null;
  channel: string;
  status: string;
  event_type: string;
  title: string;
  body: string | null;
  action_url: string | null;
  alert_id: string | null;
  opportunity_id: string | null;
  payload: Record<string, any>;
  read_at: string | null;
  sent_at: string | null;
  failed_at: string | null;
  error_message: string | null;
  attempt_count: number;
  created_at: string;
}

// ------------------------------------------------------------------- requests

export interface OpportunityDetectRequest {
  lookback_days?: number;
  min_impressions?: number;
}

export interface OpportunityStatusUpdateRequest {
  status: "open" | "in_progress" | "actioned" | "dismissed";
  dismiss_reason?: string;
}

export interface AlertStatusUpdateRequest {
  status: AlertStatus;
  resolution_note?: string;
  mute_hours?: number;
}

// ----------------------------------------------------------------- formatting

/** Integer/decimal display. Returns the em-dash placeholder for null. */
export function formatNumberFa(value: number | null | undefined): string {
  if (value === null || value === undefined || Number.isNaN(value)) return "—";
  return value.toLocaleString("fa-IR");
}

/** Backend CTR is a 0-1 float. Shown as a percentage with one decimal. */
export function formatCtrFa(ctr: number | null | undefined): string {
  if (ctr === null || ctr === undefined || Number.isNaN(ctr)) return "—";
  return `${(ctr * 100).toLocaleString("fa-IR", {
    minimumFractionDigits: 1,
    maximumFractionDigits: 1,
  })}٪`;
}

/** Average position, one decimal. */
export function formatPositionFa(position: number | null | undefined): string {
  if (position === null || position === undefined || Number.isNaN(position)) return "—";
  return position.toLocaleString("fa-IR", {
    minimumFractionDigits: 1,
    maximumFractionDigits: 1,
  });
}

/** Signed change percentage, e.g. "+۱۲.۳٪". Already a percentage, not a ratio. */
export function formatChangePercentFa(change: number | null | undefined): string {
  if (change === null || change === undefined || Number.isNaN(change)) return "—";
  const sign = change > 0 ? "+" : change < 0 ? "−" : "";
  return `${sign}${Math.abs(change).toLocaleString("fa-IR", {
    minimumFractionDigits: 1,
    maximumFractionDigits: 1,
  })}٪`;
}

export function formatDateTimeFa(iso: string | null | undefined): string {
  if (!iso) return "—";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return "—";
  return d.toLocaleDateString("fa-IR", {
    year: "numeric",
    month: "long",
    day: "numeric",
    hour: "2-digit",
    minute: "2-digit",
  });
}

export function formatDateFa(iso: string | null | undefined): string {
  if (!iso) return "—";
  const d = new Date(iso);
  if (Number.isNaN(d.getTime())) return "—";
  return d.toLocaleDateString("fa-IR");
}

/** Persian label lookup that degrades to the raw key instead of showing undefined. */
export function labelFa(map: Record<string, string>, key: string | null | undefined): string {
  if (!key) return "—";
  return map[key] ?? key;
}

// ---------------------------------------------------------------------- calls

export function getDashboardSummary() {
  return api.get<DashboardSummary>("/dashboard/summary");
}

export function listOpportunities(params: {
  website_id: string;
  status?: string;
  opportunity_type?: string;
  limit?: number;
  offset?: number;
}) {
  const qs = new URLSearchParams();
  qs.set("website_id", params.website_id);
  if (params.status) qs.set("status", params.status);
  if (params.opportunity_type) qs.set("opportunity_type", params.opportunity_type);
  if (params.limit !== undefined) qs.set("limit", String(params.limit));
  if (params.offset !== undefined) qs.set("offset", String(params.offset));
  return api.get<Opportunity[]>(`/opportunities?${qs.toString()}`);
}

export function getOpportunitySummary(websiteId: string) {
  return api.get<OpportunitySummary>(`/opportunities/summary?website_id=${websiteId}`);
}

export function detectOpportunities(websiteId: string, body?: OpportunityDetectRequest) {
  return api.post<OpportunityDetectResult>(
    `/opportunities/detect?website_id=${websiteId}`,
    body || {}
  );
}

export function updateOpportunityStatus(
  opportunityId: string,
  body: OpportunityStatusUpdateRequest
) {
  return api.patch<Opportunity>(`/opportunities/${opportunityId}/status`, body);
}

export function listAlerts(params?: {
  website_id?: string;
  status?: string;
  severity?: string;
  limit?: number;
  offset?: number;
}) {
  const qs = new URLSearchParams();
  if (params?.website_id) qs.set("website_id", params.website_id);
  if (params?.status) qs.set("status", params.status);
  if (params?.severity) qs.set("severity", params.severity);
  if (params?.limit !== undefined) qs.set("limit", String(params.limit));
  if (params?.offset !== undefined) qs.set("offset", String(params.offset));
  const qsStr = qs.toString();
  return api.get<Alert[]>(`/alerts${qsStr ? `?${qsStr}` : ""}`);
}

export function getAlertSummary() {
  return api.get<AlertSummary>("/alerts/summary");
}

export function updateAlertStatus(alertId: string, body: AlertStatusUpdateRequest) {
  return api.patch<Alert>(`/alerts/${alertId}/status`, body);
}

export function listNotifications(params?: { unread_only?: boolean; limit?: number; offset?: number }) {
  const qs = new URLSearchParams();
  if (params?.unread_only !== undefined) qs.set("unread_only", String(params.unread_only));
  if (params?.limit !== undefined) qs.set("limit", String(params.limit));
  if (params?.offset !== undefined) qs.set("offset", String(params.offset));
  const qsStr = qs.toString();
  return api.get<Notification[]>(`/notifications${qsStr ? `?${qsStr}` : ""}`);
}

export function getUnreadNotificationCount() {
  return api.get<{ unread: number }>("/notifications/unread-count");
}

export function markNotificationsRead(notificationIds?: string[]) {
  return api.post<{ marked: number }>("/notifications/mark-read", {
    notification_ids: notificationIds,
  });
}

File: backend/app/schemas/insights.py
"""Schemas for the Opportunities / Alerts / Notifications layer.

Read models mirror the ORM rows; the Numeric columns come back as Decimal from
asyncpg, so they are declared `float | None` and pydantic coerces them. Write
models are deliberately narrow: a client may only change lifecycle fields
(dismiss, acknowledge, resolve, mark read). Everything else is detector output
and must not be user-editable, otherwise the fingerprint dedup breaks.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, computed_field

from app.models.insights import (
    ALERT_SEVERITIES,
    ALERT_STATUSES,
    NOTIFICATION_CHANNELS,
    OPPORTUNITY_STATUSES,
    OPPORTUNITY_TYPES,
)


# --------------------------------------------------------------- opportunities

class OpportunityRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    website_id: UUID
    opportunity_type: str
    status: str
    title: str
    description: str | None = None
    query: str | None = None
    page_url: str | None = None
    keyword_id: UUID | None = None
    priority_score: int
    estimated_traffic_gain: int
    current_position: float | None = None
    current_clicks: int
    current_impressions: int
    current_ctr: float | None = None
    details: dict = Field(default_factory=dict)
    recommended_action: str | None = None
    detected_at: datetime
    last_seen_at: datetime | None = None
    actioned_at: datetime | None = None
    dismissed_at: datetime | None = None
    dismiss_reason: str | None = None
    linked_brief_id: UUID | None = None
    created_at: datetime
    updated_at: datetime


class OpportunityStatusUpdate(BaseModel):
    """Move an opportunity along its lifecycle."""
    status: str
    dismiss_reason: str | None = Field(default=None, max_length=1000)

    @field_validator("status")
    @classmethod
    def _known_status(cls, v: str) -> str:
        # Only human-driven transitions are accepted here. "expired" is set by
        # the detector when a finding stops reproducing, never by a client.
        allowed = {"open", "in_progress", "actioned", "dismissed"}
        if v not in allowed:
            raise ValueError(f"status must be one of {sorted(allowed)}")
        return v


class OpportunityDetectRequest(BaseModel):
    """Run the detectors for one website over a lookback window."""
    lookback_days: int = Field(default=28, ge=7, le=180)
    min_impressions: int = Field(default=1, ge=1, le=100_000)


class OpportunityDetectResult(BaseModel):
    website_id: UUID
    scanned_queries: int
    scanned_pages: int
    created: int
    updated: int
    expired: int
    by_type: dict[str, int] = Field(default_factory=dict)


class OpportunitySummary(BaseModel):
    total_open: int
    by_type: dict[str, int] = Field(default_factory=dict)
    total_estimated_traffic_gain: int = 0
    top: list[OpportunityRead] = Field(default_factory=list)


# ---------------------------------------------------------------------- alerts

class AlertRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    website_id: UUID
    alert_type: str
    severity: str
    status: str
    title: str
    message: str
    metric_name: str | None = None
    current_value: float | None = None
    previous_value: float | None = None
    change_percent: float | None = None
    entity_type: str | None = None
    entity_id: UUID | None = None
    details: dict = Field(default_factory=dict)
    occurrence_count: int
    triggered_at: datetime
    last_seen_at: datetime | None = None
    acknowledged_at: datetime | None = None
    resolved_at: datetime | None = None
    resolution_note: str | None = None
    muted_until: datetime | None = None
    notified_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class AlertStatusUpdate(BaseModel):
    status: str
    resolution_note: str | None = Field(default=None, max_length=1000)
    mute_hours: int | None = Field(default=None, ge=1, le=720)

    @field_validator("status")
    @classmethod
    def _known_status(cls, v: str) -> str:
        if v not in ALERT_STATUSES:
            raise ValueError(f"status must be one of {sorted(ALERT_STATUSES)}")
        return v


class AlertDetectRequest(BaseModel):
    """Compare the recent window against the one before it."""
    window_days: int = Field(default=7, ge=1, le=90)
    drop_threshold_percent: float = Field(default=20.0, ge=1.0, le=99.0)


class AlertDetectResult(BaseModel):
    website_id: UUID
    created: int
    updated: int
    resolved: int
    by_type: dict[str, int] = Field(default_factory=dict)
    skipped_reason: str | None = None


class AlertSummary(BaseModel):
    """Active-alert counts for the badge, one organization.

    `active` is the total and the three named severities are broken out because
    the header badge colours itself on `critical` alone without walking the map.
    `by_severity` is kept as well so a new severity added to ALERT_SEVERITIES
    still surfaces without a schema change.
    """
    active: int = 0
    critical: int = 0
    warning: int = 0
    info: int = 0
    by_severity: dict[str, int] = Field(default_factory=dict)


# --------------------------------------------------------------- notifications

class NotificationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    user_id: UUID | None = None
    website_id: UUID | None = None
    channel: str
    status: str
    event_type: str
    title: str
    body: str | None = None
    action_url: str | None = None
    alert_id: UUID | None = None
    opportunity_id: UUID | None = None
    payload: dict = Field(default_factory=dict)
    read_at: datetime | None = None
    sent_at: datetime | None = None
    failed_at: datetime | None = None
    error_message: str | None = None
    attempt_count: int
    created_at: datetime

    @computed_field
    @property
    def is_read(self) -> bool:
        return self.read_at is not None

    @computed_field
    @property
    def type(self) -> str:
        return self.event_type




class NotificationMarkReadRequest(BaseModel):
    """Empty body marks everything unread as read for the caller."""
    notification_ids: list[UUID] | None = None


class NotificationMarkReadResult(BaseModel):
    marked: int


class UnreadCountResult(BaseModel):
    unread: int


# ------------------------------------------------------------------- dashboard

class DashboardWebsiteRow(BaseModel):
    website_id: UUID
    name: str
    domain: str
    clicks: int
    impressions: int
    ctr: float
    avg_position: float
    health_score: int
    open_alerts: int
    open_opportunities: int


class DashboardSummary(BaseModel):
    """Everything the dashboard home needs in one round trip."""
    organization_id: UUID
    website_count: int
    project_count: int
    health_score: int
    total_clicks: int
    total_impressions: int
    avg_ctr: float
    avg_position: float
    clicks_change_percent: float | None = None
    impressions_change_percent: float | None = None
    active_alerts: int
    critical_alerts: int
    open_opportunities: int
    estimated_traffic_gain: int
    published_articles: int
    draft_articles: int
    last_audit_score: int | None = None
    last_gsc_sync_at: datetime | None = None
    websites: list[DashboardWebsiteRow] = Field(default_factory=list)


__all__ = [
    "OpportunityRead",
    "OpportunityStatusUpdate",
    "OpportunityDetectRequest",
    "OpportunityDetectResult",
    "OpportunitySummary",
    "AlertRead",
    "AlertStatusUpdate",
    "AlertDetectRequest",
    "AlertDetectResult",
    "AlertSummary",
    "NotificationRead",
    "NotificationMarkReadRequest",
    "NotificationMarkReadResult",
    "UnreadCountResult",
    "DashboardWebsiteRow",
    "DashboardSummary",
    "OPPORTUNITY_TYPES",
    "OPPORTUNITY_STATUSES",
    "ALERT_SEVERITIES",
    "NOTIFICATION_CHANNELS",
]
